tar/lib/src/header.dart

```diff
@@ -0,0 +1,667 @@
+        groupName == otherHeader.groupName &&
+        accessed == otherHeader.accessed &&
+        changed == otherHeader.changed &&
+        devMajor == otherHeader.devMajor &&
+        devMinor == otherHeader.devMinor &&
+        MapEquality().equals(paxRecords, otherHeader.paxRecords) &&
+        format == otherHeader.format &&
+        typeFlag == otherHeader.typeFlag;
+  }
+
+  @override
+  int get hashCode {
+    return name.hashCode ^
+        (modified.hashCode ?? 0) ^
+        linkName.hashCode ^
+        mode ^
+        size ^
+        userName.hashCode ^
+        userId ^
+        groupId ^
+        groupName.hashCode ^
+        (accessed.hashCode ?? 0) ^
+        (changed.hashCode ?? 0) ^
+        devMajor ^
+        devMinor ^
+        MapEquality().hash(paxRecords) ^
+        (format.hashCode ?? 0) ^
+        typeFlag.hashCode;
+  }
+
+  @override
+  String toString() {
+    return 'Name: $name\n'
+        'lastModifiedTime: $modified\n'
+        'linkName: $linkName\n'
+        'Mode: $mode\n'
+        'Size: $size\n'
+        'Username: $userName\n'
+        'UserId: $userId\n'
+        'GroupId: $groupId\n'
+        'GroupName: $groupName\n'
+        'AccessTime: $accessed\n'
+        'ChangeTime: $changed\n'
+        'DevMajor: $devMajor\n'
+        'DevMinor: $devMinor\n'
+        'PAXRecords: $paxRecords\n'
+        'Format: $format\n'
+        'LinkFlag: $typeFlag\n';
+  }
+}
+
+/// Checks that [rawHeader] represents a valid tar header based on the
+/// checksum, and then attempts to guess the specific format based
+/// on magic values. If the checksum fails, then an error is thrown.
+TarFormat _getFormat(List<int> rawHeader) {
+  ArgumentError.checkNotNull(rawHeader, 'rawHeader');
+
+  try {
+    final checksum = parseOctal(rawHeader, 148, 156);
+
+    /// Modern TAR archives use the unsigned checksum, but we check the signed
+    /// checksum as well for compatibility.
+    if (checksum != computeSignedCheckSum(rawHeader) &&
+        checksum != computeUnsignedCheckSum(rawHeader)) {
+      headerException('Checksum does not match');
+    }
+
+    final magic = String.fromCharCodes(rawHeader, 257, 263);
+    final version = String.fromCharCodes(rawHeader, 263, 265);
+    final trailer = String.fromCharCodes(rawHeader, 508, 512);
+    if (magic == magicUSTAR && trailer == trailerSTAR) {
+      return TarFormat.STAR;
+    }
+
+    if (magic == magicUSTAR) {
+      return TarFormat.USTAR | TarFormat.PAX;
+    }
+
+    if (magic == magicGNU && version == versionGNU) {
+      return TarFormat.GNU;
+    }
+  } catch (e) {
+    headerException('Unable to determine format of Header');
+  }
+  return TarFormat.V7;
+}
+
+/// Represents a [length]-sized fragment at [offset] in a file.
+///
+/// [SparseEntry]s can represent either data or holes, and we can easily
+/// convert between the two if we know the size of the file, all the sparse
+/// data and all the sparse entries combined must give the full size.
+class SparseEntry {
```

Comment:
Does this need to be exported? I'm just wondering...

Reply:
nope - moved it and hid it!